src/bust.py

```diff
@@ -251,21 +234,18 @@ async def play_next_song(self, skip_count: int = 0) -> None:
         await discord_utils.try_set_pin(self.now_playing_msg, True)
 
         # Called when song finishes playing
-        def ffmpeg_post_hook(e: BaseException = None):
+        async def ffmpeg_post_hook(e: Exception = None):
             if e is not None:
                 print("Song playback quit with error:", e)
-
-            # Unpin song
-            asyncio.run_coroutine_threadsafe(
-                discord_utils.try_set_pin(self.now_playing_msg, False), self.client.loop
-            )
-            self.now_playing_msg = None
-
-            # Play next song if we were not stopped
-            if not self.play_next_cancelled:
-                self.play_next_coro()
+            # Unpin now playing message
+            if self.now_playing_msg:
+                await discord_utils.try_set_pin(self.now_playing_msg, False)
+                self.now_playing_msg = None
+            await play_lock.release()
 
         # Play song
+        play_lock = asyncio.Lock()
+        await play_lock.acquire()
```

Comment:
Why is this `await play_lock.acquire()` needed when we also do the same below?

Reply:
Maybe I should write a little comment about the flow here because it might be a bit confusing with no context. The order of operations is

1. Acquire lock
2. Start playback
3. Try to acquire lock again, but get blocked because it's already acquired
4. Eventually the song ends, and it releases the lock
5. The second `acquire()` now gets unblocked, and the function returns

Without the first `acquire()`, that acquire before returning would succeed. In order for it to block until playback finishes, it needs to be acquired during the course of playback and then released when playback is done.

Reply:
Added clarifying comment
```python
# Acquire the lock during playback so that on release, play_song() returns
```